### PR TITLE
Changes to BHCal Digitization and Implementation of Adjancency Clustering

### PR DESCRIPTION
**Pull request by ruse-traveler**
### Briefly, what does this PR introduce?

This PR updates several parameters related to digitization in the BHCal to be more realistic, and implements the adjancency matrix parameter for the BHCal.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ X] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ X] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?

While it doesn't introduce breaking changes, it will result in warning messages such as those in the screenshot below.

JanaMcHitIssue

### Does this PR change default behavior?

Yes: BHCal clusters will now be clustered using the reconstructed hits collection (rather than merged hits) according to the new adjacency expression.

### CODE REVIEW

File: src/detectors/HCAL/HCAL.cc [hunk 1]
```diff
@@ -45,7 +45,6 @@ extern "C" {
 
         app->Add(new JFactoryGeneratorT<RawCalorimeterHit_factory_HcalBarrelRawHits>());
         app->Add(new JFactoryGeneratorT<CalorimeterHit_factory_HcalBarrelRecHits>());
-        app->Add(new JFactoryGeneratorT<CalorimeterHit_factory_HcalBarrelMergedHits>());
```
Comment (veprbl) [inline src/detectors/HCAL/HCAL.cc:48]: This warrants removal of a corresponding file.

File: src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h [hunk 2]
```diff
@@ -38,16 +38,16 @@ class RawCalorimeterHit_factory_HcalBarrelRawHits : public JFactoryT<edm4hep::Ra
...
-        m_resolutionTDC = 1.0 * dd4hep::ns;
+        m_resolutionTDC = 1.0 * dd4hep::picosecond;
         m_corrMeanScale = 1.0;
-        u_fields={};
-        u_refs={1, 0};
-        m_geoSvcName = "ActsGeometryProvider";
-        m_readout = "";
+        u_fields={"tile"};
+        u_refs={0};
+        m_geoSvcName = "geoServiceName";
```
Comment (veprbl) [inline src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h:49]: I assume, this change has no intended effect?

Comment (veprbl) [reply src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h:49]: (Filed #558)

File: src/detectors/HCAL/HCAL.cc (hunk 1 above)
Comment (ruse-traveler) [reply src/detectors/HCAL/HCAL.cc:48]: Ah! Would the best practice be to delete the corresponding .h file, or to leave the .h file and this line in?

File: src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h [hunk 3]
```diff
@@ -26,16 +26,19 @@ class ProtoCluster_factory_HcalBarrelIslandProtoClusters : public JFactoryT<edm4
...
         m_minClusterCenterEdep=30.0 * dd4hep::MeV; // from https://eicweb.phy.anl.gov/EIC/detectors/athena/-/blob/master/calibrations/ffi_zdc.json
 
         // adjacency matrix
+        // [Derek A.] magic numbers:
+        //   32 = number of sectors
+        //   24 = number of towers per sector in phi
         m_geoSvcName = "GeoSvc";
-        u_adjacencyMatrix = "";
-        m_readout = "";
+        u_adjacencyMatrix = "((((abs(sector_1-sector_2)==0)&&(abs(tower_1-tower_2)==1))==1)+(((abs(sector_1-sector_2)==0)&&(abs(fmod(tower_1,24)-fmod(tower_2,24))==0))==1)+(((abs(sector_1-sector_2)==1)&&(abs(fmod(tower_1,24)-fmod(tower_2,24))==0))==1)+(((abs(fmod(sector_1,31)-fmod(sector_2,31))==0)&&(abs(fmod(tower_1,24)-fmod(tower_2,24))==0))==1))==1";
```
[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h]: Whoops! Good catch! I removed the modulos from the phi-adjacency checks (terms 2 - 4), and (just to be extra safe) replaced the + with ||.

At this point, I'm pretty confident, the expression in the PR does I what I want it to, but I'm having trouble getting the SciGlass expression to work out... The modified expression for the BHCal (24 towers per phi per sector and 32 sectors) would be:
```
(
  abs(tower_1 - tower_2)
   + min(
              abs((sector_1 - sector_2) * 24 + fmod(tower_1,24) - fmod(tower_2,24)),
              64 * 24 - abs((sector_1 - sector_2) * 24 + fmod(tower_1,24) - fmod(tower_2,24))
        )
) == 1
```

[the remark this answers is not in the document]
Comment (veprbl) [reply src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h]: Your new terms, look better, apart from redundant `(cond)==1` checks. If you insist on sticking to, I could review again, but maybe adjust formatting to add newlines first.

Would that work better:
```
(
  abs(fmod(tower_1, 24) - fmod(tower_2, 24))
   + min(
              abs((sector_1 - sector_2) * 2 + fmod(tower_1, 24) - fmod(tower_2, 24)),
              32 * 2 - abs((sector_1 - sector_2) * 2 + fmod(tower_1, 24) - fmod(tower_2, 24))
        )
) == 1
```

[the remark this answers is not in the document]
Comment (veprbl) [reply src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h]: Actually,
```
(
  abs(fmod(tower_1, 24) - fmod(tower_2, 24))
   + min(
              abs((sector_1 - sector_2) * 2 + floor(tower_1 / 24) - floor(tower_2 / 24)),
              32 * 2 - abs((sector_1 - sector_2) * 2 + floor(tower_1 / 24) - floor(tower_2 / 24))
        )
) == 1
```
(I think we've discussed those substitutions on Mattermost)

File: src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h (hunk 2 above)
Comment (veprbl) [reply src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h:49]: I must admit I don't yet know what `u_fields`/`u_refs` does. My comment was regarding change to the value of `m_geoSvcName`.

File: src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (ruse-traveler) [reply src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h]: Ah! That last expression works! (We had discussed this, but I was forgetting the floors functions...)

And it looks like it does (more or less) the same thing as my clunky expression: below is the energies for BHCal-only clusters for  single10 GeV pi- where the blue points are the expression as written in the PR and the red circles are that last expression.

MyExpWithNoModsPlusLogicalORingVsSciGlassWithModsPlusFloors

There's some differences, but those could be edge cases I didn't consider in my expression. I'm good with going with the SciGlass expression!

File: src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h (hunk 2 above)
Comment (ruse-traveler) [reply src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h:49]: Oh! Yes! That change had no intended effect...

File: src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (veprbl) [reply src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h]: Looks very nice

File: src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h [hunk 4]
```diff
@@ -26,16 +26,29 @@ class ProtoCluster_factory_HcalBarrelIslandProtoClusters : public JFactoryT<edm4
...
         m_splitCluster=false;              // from https://eicweb.phy.anl.gov/EIC/detectors/athena/-/blob/master/calibrations/ffi_zdc.json
         m_minClusterHitEdep=3.0 * dd4hep::MeV;    // from https://eicweb.phy.anl.gov/EIC/detectors/athena/-/blob/master/calibrations/ffi_zdc.json
         m_minClusterCenterEdep=30.0 * dd4hep::MeV; // from https://eicweb.phy.anl.gov/EIC/detectors/athena/-/blob/master/calibrations/ffi_zdc.json
 
         // adjacency matrix
         m_geoSvcName = "GeoSvc";
-        u_adjacencyMatrix = "";
-        m_readout = "";
+        // Magic constants:
+        //   32 - number of sectors
```
Comment (veprbl) [inline src/detectors/HCAL/ProtoCluster_factory_HcalBarrelIslandProtoClusters.h:38]:
```suggestion
        //   32 - number of sectors
        //   2 - number of rows per sector
```

File: src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h [hunk 5]
```diff
@@ -38,16 +38,16 @@ class RawCalorimeterHit_factory_HcalBarrelRawHits : public JFactoryT<edm4hep::Ra
...
         m_pedSigmaADC = 2.0;
-        m_resolutionTDC = 1.0 * dd4hep::ns;
+        m_resolutionTDC = 1.0 * dd4hep::picosecond;
         m_corrMeanScale = 1.0;
-        u_fields={};
-        u_refs={1, 0};
-        m_geoSvcName = "ActsGeometryProvider";
-        m_readout = "";
+        u_fields={"tile"};
+        u_refs={0};
```
[the remark this answers is not in the document]
Comment (veprbl) [reply src/detectors/HCAL/RawCalorimeterHit_factory_HcalBarrelRawHits.h:48]: Aha! Thant makes sense now. Thank you.

File: src/detectors/HCAL/HCAL.cc (hunk 1 above)
Comment (veprbl) [reply src/detectors/HCAL/HCAL.cc:48]: Yes, delete the .h and .cc files with the corresponding factory.

Normally, you'd need to also remove a `"HcalBarrelMergedHits"` line in
https://github.com/eic/EICrecon/blob/1f30d3b5e2aef6fdc9e7bccfe8b4e1228f63b551/src/services/io/podio/JEventProcessorPODIO.cc#L167-L179
but it seems like they forgot to add it in the first place.

Comment (ruse-traveler) [reply src/detectors/HCAL/HCAL.cc:48]: Gotcha! I'll remove those!

**Review by veprbl — APPROVED** [review]

**Pull request merged**